src/cowboy_http2.erl

```diff
@@ -938,6 +964,54 @@ goaway_streams(State, [{StreamID, #stream{state=StreamState}}|Tail], LastStreamI
 goaway_streams(State, [Stream|Tail], LastStreamID, Reason, Acc) ->
 	goaway_streams(State, Tail, LastStreamID, Reason, [Stream|Acc]).
 
+%% A server that is attempting to gracefully shut down a connection SHOULD send
+%% an initial GOAWAY frame with the last stream identifier set to 2^31-1 and a
+%% NO_ERROR code. This signals to the client that a shutdown is imminent and
+%% that initiating further requests is prohibited. After allowing time for any
+%% in-flight stream creation (at least one round-trip time), the server can send
+%% another GOAWAY frame with an updated last stream identifier. This ensures
+%% that a connection can be cleanly shut down without losing requests.
+-spec graceful_shutdown(#state{}, _) -> #state{}.
+graceful_shutdown(State=#state{socket=Socket, transport=Transport, http2_status=Status}, Reason)
+		when Status =:= connected; Status =:= closing ->
+	%% @todo Save the timer ref in the state?
+	_TRef = erlang:start_timer(?shutdown_rtt, self(), {graceful_shutdown, Reason}),
+	case Status of
+		connected ->
+			Transport:send(Socket, cow_http2:goaway(
+				16#7FFFFFFF, %% 2^31 - 1
+				no_error, <<>>));
+		closing ->
+			ok
+	end,
+	State;
+graceful_shutdown(State, Reason) ->
+	terminate(State, {stop, Reason, 'The connection is going away.'}).
+
+%% 1 RTT after sending the initial GOAWAY. Any in-flight requests have been
+%% received. Initiate 'closing' state and don't accept any new streams.
+-spec graceful_shutdown_continue(#state{}, Reason :: term()) -> #state{}.
+graceful_shutdown_continue(State=#state{streams=Streams}, _Reason) when Streams =:= #{} ->
+	terminate(State, {stop, normal, 'The connection is going away.'});
+graceful_shutdown_continue(State=#state{http2_status=connected, http2_machine=HTTP2Machine,
+		socket=Socket, transport=Transport}, Reason) ->
+	%% Stop accepting new streams.
+	Transport:send(Socket, cow_http2:goaway(
+		cow_http2_machine:get_last_streamid(HTTP2Machine),
+		no_error, <<>>)),
+	graceful_shutdown_continue(State#state{http2_status=closing}, Reason);
+graceful_shutdown_continue(State=#state{http2_status=closing, opts=Opts}, Reason) ->
+	%% Give ongoing streams some time to complete.
+	case maps:get(shutdown_timeout, Opts, 5000) of
+		infinity ->
+			State; %% No hard shutdown. All streams have to finish gracefully.
+		ShutdownTimeout ->
+			%% Shutdown hard just before the supervisor kills us.
+			Timeout = max(ShutdownTimeout - ?shutdown_rtt - 100, 0),
+			_ = erlang:start_timer(Timeout, self(), {hard_shutdown, Reason}),
```

Comment:
Don't do calculations like this. You want one configurable value for the time until we stop accepting new streams during shutdown, and one configurable value for the time to shutdown itself.

Reply:
OK! (I was also thinking about calculating a default in cowboy:start_clear and start_tls by stealing the ranch shutdown and injecting it in the protocol options to make sure we shut down in time before we're killed, but I guess that's out of the question... Fine!)

This means we can't reuse the already existing `shutdown_timeout` configuration which defaults to 5000, since we would be killed by the supervisor before the timeout fires off, right?

How about a new option `init_shutdown_timeout` (default 500) and changing the default for `shutdown_timeout` to say 4000? Or would you prefer adding another option instead of changing the default of the existing one?

Reply:
`shutdown_timeout` is about killing child processes, not about the connection shutdown. It's not dependent on the protocol. The value it has is fine.

The `shutdown` option in Ranch is the one that's relevant here, and it defaults to 5000. So the total graceful shutdown should by default take less time. 1000+3000 for example would be plenty of time.

I think that since it's something specific to HTTP/2, maybe we should name the two new options `goaway_something_timeout`. Maybe `goaway_closing_timeout` and `goaway_close_timeout`. I don't know, naming is hard.

Reply:
"closing" and "close" can cause confusion I think. How about `goaway_initial_timeout` and `goaway_close_timeout`? (The RFC refers to "an initial GOAWAY frame".)

Reply:
I don't think it's much better but let's go with this for now. Let's see how it looks like once it has a documentation entry explaining what it's about.

Reply:
OK, I'll go with that for now.

Just a thought: For HTTP/1.1, graceful shutdown can be accomplished by sending `connection: close` after finishing the current request, with a timeout for forcing it, so maybe at least one of the options could be useful outside of HTTP/2.

Reply:
For HTTP/1.1 besides marking the current stream as being the last, I don't think there's any need for a timeout in the supervisor shutdown case. The supervisor can kill the process. Only the `sys:terminate` case would need a timeout.

Reply:
True. And for `sys:terminate`, if it doesn't finish soon enough, you could always `exit(Conn, kill)` the process, so perhaps there is no need for a timer for HTTP/1.